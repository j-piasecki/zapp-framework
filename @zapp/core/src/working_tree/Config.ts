export class Config {
  public id: string

  constructor(id: string) {
    this.id = id
  }
}
