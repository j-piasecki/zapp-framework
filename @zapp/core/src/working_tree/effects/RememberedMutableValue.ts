import { RememberNode } from '../RememberNode.js'
import { WorkingTree } from '../WorkingTree.js'
import { RememberedValue } from './RememberedValue.js'

export class RememberedMutableValue<T> extends RememberedValue<T> {
  public get value() {
    return this._value
  }

  public set value(val: T) {
    this._value = val

    if (this.context.isDropped) {
      const newNode = WorkingTree.root.getNodeFromPath(this.context.path.concat(this.context.id))
      if (newNode !== null) {
        const remembered = (newNode as RememberNode).remembered as RememberedMutableValue<T>
        remembered.value = val
      }
    } else {
      WorkingTree.queueUpdate(this.context.parent!)
    }
  }
}
