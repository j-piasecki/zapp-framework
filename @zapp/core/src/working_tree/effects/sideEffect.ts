import { EffectNode } from '../EffectNode.js'
import { ViewNode } from '../ViewNode.js'
import { WorkingTree } from '../WorkingTree.js'

export function sideEffect(effect: (isRestoring: boolean) => (() => void) | void, ...keys: any) {
  const current = WorkingTree.current as ViewNode
  const context = current.effect()

  const path = context.path.slice(current.rememberedContext?.path.length).concat(context.id)
  const rememberedNode = current.rememberedContext?.getNodeFromPath(path) ?? null

  if (rememberedNode instanceof EffectNode) {
    let relaunch = false

    if (keys.length !== rememberedNode.keys.length) {
      relaunch = true
    } else {
      for (let i = 0; i < keys.length; i++) {
        if (keys[i] !== rememberedNode.keys[i]) {
          relaunch = true
          break
        }
      }
    }

    if (relaunch) {
      rememberedNode.effectCleanup?.()

      context.effect = effect
      WorkingTree.withContext(context, () => {
        context.effectCleanup = effect(WorkingTree.isRestoringState()) as (() => void) | undefined
      })
      context.keys = keys
    } else {
      context.effect = rememberedNode.effect
      context.effectCleanup = rememberedNode.effectCleanup
      context.keys = rememberedNode.keys
    }
  } else {
    context.effect = effect
    WorkingTree.withContext(context, () => {
      context.effectCleanup = effect(WorkingTree.isRestoringState()) as (() => void) | undefined
    })
    context.keys = keys
  }

  current.children.push(context)
}
