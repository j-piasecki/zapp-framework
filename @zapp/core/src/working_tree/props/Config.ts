export function Config(id: string) {
  return new ConfigBuilder(id)
}

export interface Padding {
  top: number
  bottom: number
  start: number
  end: number
}

export interface ConfigType {
  id: string

  fillSize?: boolean

  weight?: number
  background?: number
  cornerRadius?: number

  width?: number
  fillWidth?: number

  height?: number
  fillHeight?: number

  padding?: Padding

  text?: string

  // props that may be inherited from the parent
  textSize?: number
  textColor?: number
}

export class ConfigBuilder {
  private config: ConfigType

  constructor(id: string) {
    this.config = {
      id: id,
    }
  }

  public build() {
    return this.config
  }

  public fillSize(): Omit<this, 'width' | 'fillWidth' | 'height' | 'fillHeight' | 'fillSize'> {
    this.config.fillSize = true
    return this
  }

  public fillWidth(portion: number): Omit<this, 'width' | 'fillWidth' | 'fillSize'> {
    this.config.fillWidth = portion
    return this
  }

  public fillHeight(portion: number): Omit<this, 'height' | 'fillHeight' | 'fillSize'> {
    this.config.fillHeight = portion
    return this
  }

  public width(width: number): Omit<this, 'width' | 'fillWidth' | 'fillSize'> {
    this.config.width = width
    return this
  }

  public height(height: number): Omit<this, 'height' | 'fillHeight' | 'fillSize'> {
    this.config.height = height
    return this
  }

  public weight(weight: number): Omit<this, 'weight'> {
    this.config.weight = weight
    return this
  }

  public padding(padding: number): Omit<this, 'padding'>
  public padding(vertical: number, horizontal: number): Omit<this, 'padding'>
  public padding(start: number, top: number, end: number, bottom: number): Omit<this, 'padding'>
  public padding(start: number, top?: number, end?: number, bottom?: number): Omit<this, 'padding'> {
    if (top !== undefined && end !== undefined && bottom !== undefined) {
      this.config.padding = {
        start: start,
        top: top,
        end: end,
        bottom: bottom,
      }
    } else if (top !== undefined) {
      this.config.padding = {
        start: top,
        top: start,
        end: top,
        bottom: start,
      }
    } else {
      this.config.padding = {
        start: start,
        top: start,
        end: start,
        bottom: start,
      }
    }

    return this
  }

  public background(background: number): Omit<this, 'background'> {
    this.config.background = background
    return this
  }

  public cornerRadius(radius: number): Omit<this, 'cornerRadius'> {
    this.config.cornerRadius = radius
    return this
  }

  public textColor(textColor: number): Omit<this, 'textColor'> {
    this.config.textColor = textColor
    return this
  }

  public textSize(textSize: number): Omit<this, 'cornerRatextSizedius'> {
    this.config.textSize = textSize
    return this
  }
}
