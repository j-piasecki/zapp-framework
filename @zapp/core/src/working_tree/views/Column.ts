import { ViewNode } from '../ViewNode.js'
import { WorkingTree } from '../WorkingTree.js'
import { NodeType } from '../../NodeType.js'

export function Column(config: any, body: () => void) {
  const current = WorkingTree.current as ViewNode

  const context = current.create({
    id: config.id,
    type: NodeType.Column,
    config: config,
    body: body,
  })

  current.children.push(context)

  WorkingTree.withContext(context, body)
}
