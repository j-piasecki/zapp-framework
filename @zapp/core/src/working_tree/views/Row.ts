import { ViewNode } from '../ViewNode.js'
import { WorkingTree } from '../WorkingTree.js'
import { NodeType } from '../../NodeType.js'

export function Row(config: any, body: () => void) {
  const current = WorkingTree.current as ViewNode

  const context = current.create({
    id: config.id,
    type: NodeType.Row,
    config: config,
    body: body,
  })

  current.children.push(context)

  WorkingTree.withContext(context, body)
}
