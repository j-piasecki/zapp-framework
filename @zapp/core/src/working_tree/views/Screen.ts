import type { RequireSome } from '../../utils.js'
import { ViewNode } from '../ViewNode.js'
import { WorkingTree } from '../WorkingTree.js'
import { NodeType } from '../../NodeType.js'
import { ConfigBuilder } from '../props/Config.js'

export function ScreenBody(
  configBuilder: RequireSome<ConfigBuilder, 'build'>,
  body?: (params?: Record<string, unknown>) => void
) {
  const config = configBuilder.build()
  if (config.background === undefined) {
    config.background = 0x000000
  }

  const current = WorkingTree.current as ViewNode

  const context = current.create({
    id: config.id,
    type: NodeType.Screen,
    config: config,
    body: body,
  })

  current.children.push(context)

  WorkingTree.withContext(context, body)
}

let simpleScreenImplementation = (
  configBuilder: RequireSome<ConfigBuilder, 'build'>,
  body?: (params?: Record<string, unknown>) => void
) => {
  ScreenBody(configBuilder, body)
}

export function setSimpleScreenImplementation(
  implementation: (
    config: RequireSome<ConfigBuilder, 'build'>,
    body?: (params?: Record<string, unknown>) => void
  ) => void
) {
  simpleScreenImplementation = implementation
}

export function SimpleScreen(
  configBuilder: RequireSome<ConfigBuilder, 'build'>,
  body?: (params?: Record<string, unknown>) => void
) {
  simpleScreenImplementation(configBuilder, body)
}
