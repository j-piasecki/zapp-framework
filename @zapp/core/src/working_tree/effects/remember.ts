import { findRelativePath } from '../../utils.js'
import { RememberNode } from '../RememberNode.js'
import { ViewNode } from '../ViewNode.js'
import { WorkingTree } from '../WorkingTree.js'
import { RememberedMutableValue } from './RememberedMutableValue.js'
import { Animation, AnimationType } from './animation/Animation.js'
import { TimingAnimation } from './animation/TimingAnimation.js'

export function remember<T>(value: T): RememberedMutableValue<T> {
  const current = WorkingTree.current as ViewNode
  const context = current.remember()

  let savedRemembered: RememberedMutableValue<T> | null = null

  const restoredState = WorkingTree.root.savedState?.tryFindingValue?.(context)

  if (restoredState !== undefined) {
    value = restoredState.value as T
  } else {
    const path = findRelativePath(context.path, current.rememberedContext?.path)?.concat(context.id)
    if (path !== null && path !== undefined) {
      const rememberedNode = current.rememberedContext?.getNodeFromPath(path) ?? null

      if (rememberedNode instanceof RememberNode && rememberedNode.remembered instanceof RememberedMutableValue) {
        savedRemembered = rememberedNode.remembered
        // TODO: investigate context switching in remembered values more
        savedRemembered.switchContext(context)
      }
    }
  }

  const result = savedRemembered === null ? new RememberedMutableValue(value, context) : savedRemembered

  if (restoredState !== undefined && restoredState.animationData !== undefined) {
    let anim: Animation<unknown> | null = null

    if (restoredState.animationData.type === AnimationType.Timing) {
      anim = TimingAnimation.restore(restoredState.animationData)
    }

    if (anim !== null) {
      // when restoring animation from saved state, assign directly
      anim.rememberedValue = result
      result.animation = anim as Animation<T>
    }
  }

  context.remembered = result
  current.children.push(context)

  return result
}
