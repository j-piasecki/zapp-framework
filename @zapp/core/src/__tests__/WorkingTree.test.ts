import { WorkingTree } from '../working_tree/WorkingTree'
import { Config } from '../working_tree/props/Config'
import { Column } from '../working_tree/views/Column'
import { Screen } from '../working_tree/views/Screen'
import { Row } from '../working_tree/views/Row'
import { RememberedMutableValue } from '../working_tree/effects/RememberedMutableValue'
import { remember } from '../working_tree/effects/remember'
import { sideEffect } from '../working_tree/effects/sideEffect'

jest.useFakeTimers()

afterEach(() => {
  WorkingTree.dropAll()
})

test('Tree is generated correctly', () => {
  Screen(Config('screen'), () => {
    Column(Config('column'), () => {
      const visible = remember(true)

      if (visible.value) {
        Row(Config('row1'), () => {})
      }
      Row(Config('row2'), () => {})
    })
  })

  const generatedTree = WorkingTree.root.toString()
  expect(generatedTree).toMatchSnapshot()
})

test('Tree is generated correctly after update', () => {
  let remembered: RememberedMutableValue<boolean>

  Screen(Config('screen'), () => {
    Column(Config('column'), () => {
      const visible = remember(true)
      remembered = visible

      if (visible.value) {
        Row(Config('row1'), () => {})
      }
      Row(Config('row2'), () => {})
    })
  })

  remembered!.value = false
  WorkingTree.performUpdate()

  const generatedTree = WorkingTree.root.toString()
  expect(generatedTree).toMatchSnapshot()
})

test('Tree is not re-generated after setting the same value', () => {
  const fn = jest.fn()
  let remembered: RememberedMutableValue<boolean>

  Screen(Config('screen'), () => {
    Column(Config('column'), () => {
      const visible = remember(true)
      remembered = visible

      fn()

      if (visible.value) {
        Row(Config('row1'), () => {})
      }
      Row(Config('row2'), () => {})
    })
  })

  remembered!.value = true
  WorkingTree.performUpdate()

  expect(fn).toBeCalledTimes(1)
})

test('Side effect is not triggered when non-key value changes', () => {
  const effect = jest.fn()
  let remembered: RememberedMutableValue<boolean>

  Screen(Config('screen'), () => {
    const value = remember(true)
    remembered = value

    Column(Config('column'), () => {
      sideEffect(() => {
        effect()
      })
    })
  })

  remembered!.value = false
  WorkingTree.performUpdate()

  expect(effect).toBeCalledTimes(1)
})

test('Side effect is triggered when key value changes', () => {
  const effect = jest.fn()
  const effectDispose = jest.fn()
  let remembered: RememberedMutableValue<number>

  Screen(Config('screen'), () => {
    const value = remember(1)
    remembered = value

    Column(Config('column'), () => {
      sideEffect(() => {
        effect()

        return effectDispose
      }, value.value)
    })
  })

  remembered!.value++
  WorkingTree.performUpdate()

  expect(effect).toBeCalledTimes(2)
  expect(effectDispose).toBeCalledTimes(1)
})

test('Side effect is disposed of when it leaves the tree', () => {
  const effect = jest.fn()
  const effectDispose = jest.fn()
  let remembered: RememberedMutableValue<boolean>

  Screen(Config('screen'), () => {
    const value = remember(true)
    remembered = value

    if (value.value) {
      Column(Config('column'), () => {
        sideEffect(() => {
          effect()

          return effectDispose
        })
      })
    }
  })

  remembered!.value = false
  WorkingTree.performUpdate()

  expect(effect).toBeCalledTimes(1)
  expect(effectDispose).toBeCalledTimes(1)
})

test('Value is updated correctly when captured in closure', () => {
  let remembered: RememberedMutableValue<number>

  Screen(Config('screen'), () => {
    const value = remember(1)
    remembered = value

    Column(Config('column'), () => {
      sideEffect(() => {
        setTimeout(() => {
          value.value = 5
          WorkingTree.performUpdate()
        }, 100)
      })
    })
  })

  remembered!.value = 2
  WorkingTree.performUpdate()

  const generatedTree1 = WorkingTree.root.toString()
  expect(generatedTree1).toMatchSnapshot()

  jest.advanceTimersByTime(100)

  const generatedTree2 = WorkingTree.root.toString()
  expect(generatedTree2).toMatchSnapshot()
})
