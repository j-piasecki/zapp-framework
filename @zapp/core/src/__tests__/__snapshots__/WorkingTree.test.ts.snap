// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree is generated correctly 1`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"parent\\": \\"no parent\\",
  \\"path\\": \\"\\",
  \\"isDropped\\": false,
  \\"body\\": \\"body\\",
  \\"children\\": [
    {
      \\"id\\": \\"screen\\",
      \\"type\\": \\"screen\\",
      \\"parent\\": \\"#__root\\",
      \\"path\\": \\"#__root\\",
      \\"isDropped\\": false,
      \\"body\\": \\"body\\",
      \\"children\\": [
        {
          \\"id\\": \\"column\\",
          \\"type\\": \\"column\\",
          \\"parent\\": \\"screen\\",
          \\"path\\": \\"#__root/screen\\",
          \\"isDropped\\": false,
          \\"body\\": \\"body\\",
          \\"children\\": [
            {
              \\"id\\": \\"0\\",
              \\"type\\": \\"remember\\",
              \\"parent\\": \\"column\\",
              \\"path\\": \\"#__root/screen/column\\",
              \\"isDropped\\": false,
              \\"remembered\\": true
            },
            {
              \\"id\\": \\"row1\\",
              \\"type\\": \\"row\\",
              \\"parent\\": \\"column\\",
              \\"path\\": \\"#__root/screen/column\\",
              \\"isDropped\\": false,
              \\"body\\": \\"body\\",
              \\"children\\": [],
              \\"config\\": {
                \\"id\\": \\"row1\\"
              },
              \\"nextActionId\\": 0,
              \\"rememberedContext\\": \\"no remembered context\\",
              \\"override\\": \\"no override\\"
            },
            {
              \\"id\\": \\"row2\\",
              \\"type\\": \\"row\\",
              \\"parent\\": \\"column\\",
              \\"path\\": \\"#__root/screen/column\\",
              \\"isDropped\\": false,
              \\"body\\": \\"body\\",
              \\"children\\": [],
              \\"config\\": {
                \\"id\\": \\"row2\\"
              },
              \\"nextActionId\\": 0,
              \\"rememberedContext\\": \\"no remembered context\\",
              \\"override\\": \\"no override\\"
            }
          ],
          \\"config\\": {
            \\"id\\": \\"column\\"
          },
          \\"nextActionId\\": 1,
          \\"rememberedContext\\": \\"no remembered context\\",
          \\"override\\": \\"no override\\"
        }
      ],
      \\"config\\": {
        \\"id\\": \\"screen\\"
      },
      \\"nextActionId\\": 0,
      \\"rememberedContext\\": \\"no remembered context\\",
      \\"override\\": \\"no override\\"
    }
  ],
  \\"config\\": {
    \\"id\\": \\"#__root\\"
  },
  \\"nextActionId\\": 0
}"
`;

exports[`Tree is generated correctly after update 1`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"parent\\": \\"no parent\\",
  \\"path\\": \\"\\",
  \\"isDropped\\": false,
  \\"body\\": \\"body\\",
  \\"children\\": [
    {
      \\"id\\": \\"screen\\",
      \\"type\\": \\"screen\\",
      \\"parent\\": \\"#__root\\",
      \\"path\\": \\"#__root\\",
      \\"isDropped\\": false,
      \\"body\\": \\"body\\",
      \\"children\\": [
        {
          \\"id\\": \\"column\\",
          \\"type\\": \\"column\\",
          \\"parent\\": \\"screen\\",
          \\"path\\": \\"#__root/screen\\",
          \\"isDropped\\": false,
          \\"body\\": \\"body\\",
          \\"children\\": [
            {
              \\"id\\": \\"0\\",
              \\"type\\": \\"remember\\",
              \\"parent\\": \\"column\\",
              \\"path\\": \\"#__root/screen/column\\",
              \\"isDropped\\": false,
              \\"remembered\\": false
            },
            {
              \\"id\\": \\"row2\\",
              \\"type\\": \\"row\\",
              \\"parent\\": \\"column\\",
              \\"path\\": \\"#__root/screen/column\\",
              \\"isDropped\\": false,
              \\"body\\": \\"body\\",
              \\"children\\": [],
              \\"config\\": {
                \\"id\\": \\"row2\\"
              },
              \\"nextActionId\\": 0,
              \\"rememberedContext\\": \\"no remembered context\\",
              \\"override\\": \\"no override\\"
            }
          ],
          \\"config\\": {
            \\"id\\": \\"column\\"
          },
          \\"nextActionId\\": 1,
          \\"rememberedContext\\": \\"no remembered context\\",
          \\"override\\": \\"no override\\"
        }
      ],
      \\"config\\": {
        \\"id\\": \\"screen\\"
      },
      \\"nextActionId\\": 0,
      \\"rememberedContext\\": \\"no remembered context\\",
      \\"override\\": \\"no override\\"
    }
  ],
  \\"config\\": {
    \\"id\\": \\"#__root\\"
  },
  \\"nextActionId\\": 0
}"
`;

exports[`Value is updated correctly when captured in closure 1`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"parent\\": \\"no parent\\",
  \\"path\\": \\"\\",
  \\"isDropped\\": false,
  \\"body\\": \\"body\\",
  \\"children\\": [
    {
      \\"id\\": \\"screen\\",
      \\"type\\": \\"screen\\",
      \\"parent\\": \\"#__root\\",
      \\"path\\": \\"#__root\\",
      \\"isDropped\\": false,
      \\"body\\": \\"body\\",
      \\"children\\": [
        {
          \\"id\\": \\"0\\",
          \\"type\\": \\"remember\\",
          \\"parent\\": \\"screen\\",
          \\"path\\": \\"#__root/screen\\",
          \\"isDropped\\": false,
          \\"remembered\\": 2
        },
        {
          \\"id\\": \\"column\\",
          \\"type\\": \\"column\\",
          \\"parent\\": \\"screen\\",
          \\"path\\": \\"#__root/screen\\",
          \\"isDropped\\": false,
          \\"body\\": \\"body\\",
          \\"children\\": [
            {
              \\"id\\": \\"0\\",
              \\"type\\": \\"effect\\",
              \\"parent\\": \\"column\\",
              \\"path\\": \\"#__root/screen/column\\",
              \\"isDropped\\": false,
              \\"effectCleanup\\": \\"no cleanup\\",
              \\"keys\\": []
            }
          ],
          \\"config\\": {
            \\"id\\": \\"column\\"
          },
          \\"nextActionId\\": 1,
          \\"rememberedContext\\": \\"no remembered context\\",
          \\"override\\": \\"no override\\"
        }
      ],
      \\"config\\": {
        \\"id\\": \\"screen\\"
      },
      \\"nextActionId\\": 1,
      \\"rememberedContext\\": \\"no remembered context\\",
      \\"override\\": \\"no override\\"
    }
  ],
  \\"config\\": {
    \\"id\\": \\"#__root\\"
  },
  \\"nextActionId\\": 0
}"
`;

exports[`Value is updated correctly when captured in closure 2`] = `
"{
  \\"id\\": \\"#__root\\",
  \\"type\\": \\"root\\",
  \\"parent\\": \\"no parent\\",
  \\"path\\": \\"\\",
  \\"isDropped\\": false,
  \\"body\\": \\"body\\",
  \\"children\\": [
    {
      \\"id\\": \\"screen\\",
      \\"type\\": \\"screen\\",
      \\"parent\\": \\"#__root\\",
      \\"path\\": \\"#__root\\",
      \\"isDropped\\": false,
      \\"body\\": \\"body\\",
      \\"children\\": [
        {
          \\"id\\": \\"0\\",
          \\"type\\": \\"remember\\",
          \\"parent\\": \\"screen\\",
          \\"path\\": \\"#__root/screen\\",
          \\"isDropped\\": false,
          \\"remembered\\": 5
        },
        {
          \\"id\\": \\"column\\",
          \\"type\\": \\"column\\",
          \\"parent\\": \\"screen\\",
          \\"path\\": \\"#__root/screen\\",
          \\"isDropped\\": false,
          \\"body\\": \\"body\\",
          \\"children\\": [
            {
              \\"id\\": \\"0\\",
              \\"type\\": \\"effect\\",
              \\"parent\\": \\"column\\",
              \\"path\\": \\"#__root/screen/column\\",
              \\"isDropped\\": false,
              \\"effectCleanup\\": \\"no cleanup\\",
              \\"keys\\": []
            }
          ],
          \\"config\\": {
            \\"id\\": \\"column\\"
          },
          \\"nextActionId\\": 1,
          \\"rememberedContext\\": \\"no remembered context\\",
          \\"override\\": \\"no override\\"
        }
      ],
      \\"config\\": {
        \\"id\\": \\"screen\\"
      },
      \\"nextActionId\\": 1,
      \\"rememberedContext\\": \\"no remembered context\\",
      \\"override\\": \\"no override\\"
    }
  ],
  \\"config\\": {
    \\"id\\": \\"#__root\\"
  },
  \\"nextActionId\\": 0
}"
`;
